src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1404,163 +1390,341 @@ object Desugar {
+            * In the case where the value variable 'v' is missing all the code annotated with [v]
             * is omitted
             */
-          case n@PAssForRange(range, ass, spec, body) =>
+          case n@PShortForRange(range, shorts, _, spec, body) if underlyingType(info.typ(range.exp)).isInstanceOf[MapT] =>
+            // is a block as it introduces new variables
             unit(block(for {
               exp <- goE(range.exp)
-              typ = underlyingType(exp.typ) match {
-                case slice : in.SliceT => slice
-                case array : in.ArrayT => array
-                case _ : in.MapT => violation("Maps are not supported yet in range")
-                case _ : in.StringT => violation("Strings are not supported yet in range")
-                case t => violation(s"Range not applicable to type $t")
-              }
-              indexLeft <- goL(ass(0))
-              rangeSrc = meta(range, info)
-              indexSrc = meta(ass(0), info)
-              rangeExpSrc = meta(range.exp, info)
 
-              copiedVar <- freshDeclaredExclusiveVar(exp.typ, n, info)(rangeExpSrc)
-              copyAss = singleAss(in.Assignee.Var(copiedVar), exp)(rangeExpSrc)
-              length = in.Length(copiedVar)(rangeSrc)
+              keyType = exp.typ.asInstanceOf[in.MapT].keys.withAddressability(Addressability.exclusiveVariable)
+              valType = exp.typ.asInstanceOf[in.MapT].values.withAddressability(Addressability.exclusiveVariable)
+              visType = in.SetT(keyType, Addressability.exclusiveVariable)
 
-              copiedIndexVar <- freshDeclaredExclusiveVar(in.IntT(Addressability.exclusiveVariable), n, info)(rangeExpSrc)
-              copyIndexAss = singleAss(in.Assignee.Var(copiedIndexVar), in.IntLit(0)(indexSrc))(rangeExpSrc)
-              incrCopiedIndex = singleAss(in.Assignee.Var(copiedIndexVar), in.Add(copiedIndexVar, in.IntLit(1)(indexSrc))(indexSrc))(indexSrc)
+              domain = in.MapKeys(exp, underlyingType(exp.typ))(src)
 
-              indexAss = in.If(in.LessCmp(in.IntLit(0)(indexSrc), length)(indexSrc), singleAss(indexLeft, in.IntLit(0)(rangeSrc))(indexSrc), in.Seqn(Vector())(indexSrc))(indexSrc)
-              incrIndex = singleAss(indexLeft, copiedIndexVar)(indexSrc)
+              visited <- leftOfAssignmentD(range.enumerated, info)(in.SetT(keyType, Addressability.exclusiveVariable))
 
-              cond = in.LessCmp(indexLeft.op, length)(rangeSrc)
+              perm <- freshDeclaredExclusiveVar(in.PermissionT(Addressability.exclusiveVariable), n, info)(src)
 
-              copiedIndexEqualsIndexInv = in.Implication(
-                in.LessCmp(in.IntLit(0)(indexSrc), length)(indexSrc),
-                in.ExprAssertion(in.EqCmp(indexLeft.op, copiedIndexVar)(rangeSrc))(rangeSrc))(rangeSrc)
+              initPerm = singleAss(in.Assignee.Var(perm), in.PermLit(1, 1000000)(src))(src)
```

Comment:
Is this hardcoded in other places too? If so, I would move it to a constant to make it easier to change